app/services/solr_mapper.rb

```diff
@@ -14,15 +14,85 @@ def initialize(metadata:, dataset_record_id:, dataset_record_set_id:)
+  # rubocop:disable Metrics/AbcSize
   def call
     {
       id: dataset_record_id,
-      dataset_record_set_id: dataset_record_set_id,
-      title: metadata.dig(:titles, 0, :title)
-    }
+      dataset_record_set_id_ss: dataset_record_set_id,
+      access_ssi: metadata['access'],
+      provider_ssi: metadata['provider'],
+      creators_struct_ss: metadata['creators'].to_json,
+      descriptions_tsim: retrieve_descriptions(metadata['descriptions'] || []),
+      doi_ssi: retrieve_doi(metadata['identifiers']) || '',
+      provider_identifier_ssim: map_provider_identifiers(metadata['identifiers'], metadata['provider'])
+    }.merge(map_titles(metadata['titles']))
   end
+  # rubocop:enable Metrics/AbcSize
 
   private
 
   attr_reader :metadata, :dataset_record_id, :dataset_record_set_id
+
+  # Given titles from metadata, return field value based on title type
+  def title_fields(title_type, titles)
+    titles.select { |title_obj| title_obj['title_type'] == title_type }.pluck('title')
+  end
+
+  def map_titles(titles_metadata)
+    titles = titles_metadata.reject { |title_obj| title_obj.key?('title_type') }.pluck('title')
+    subtitles = title_fields('Subtitle', titles_metadata)
+    alternative_titles = title_fields('AlternativeTitle', titles_metadata)
+    translated_titles =  title_fields('TranslatedTitle', titles_metadata)
+    other_titles = title_fields('Other', titles_metadata)
+
+    # The schema requires that a title of any type be present
+    # The Solr mapping requires a dedicated title field, so this ensures
+    # the main title field is populated
+    titles = titles.presence || subtitles.presence || alternative_titles.presence ||
+             translated_titles.presence || other_titles.presence
+
+    map_typed_title_fields(titles, subtitles, alternative_titles, translated_titles, other_titles)
+  end
+
+  def map_typed_title_fields(titles, subtitles, alternative_titles, translated_titles, other_titles)
+    # If one of these fields is an empty array, Solr is setup to not add that field
+    # so we don't have to check if any of the title arrays or empty or not
+    {}.tap do |mapped_titles|
+      mapped_titles[:title_tsim] = titles
+      mapped_titles[:subtitle_tsim] = subtitles
+      mapped_titles[:alternative_title_tsim] = alternative_titles
+      mapped_titles[:translate_title_tsim] = translated_titles
+      mapped_titles[:other_title_tsim] = other_titles
+    end
+  end
+
+  def map_provider_identifiers(identifiers_metadata, provider)
+    matching_provider_ids(identifiers_metadata, provider).pluck('identifier')
+  end
+
+  def matching_provider_ids(identifiers_metadata, provider)
+    identifiers_metadata.select { |id_info| id_info['identifier_type'] == provider_ref(provider) }
+  end
+
+  def provider_ref(provider)
+    case provider
+    when 'DataCite'
+      'DOI'
+    when 'Zenodo'
+      'ZenodoId'
+    else
+      "#{provider}Reference"
+    end
+  end
+
+  def retrieve_doi(identifiers_metadata)
```

Comment:
`identifiers_metadata` doesn't need to be passed as a variable since `metadata` is an instance variable.

Can you call this method `doi_field`? `retrieve_doi` is confusing.

How about something like this?
```
metadata['identifiers'].find {|id_hash| id_hash['identifier_type'] == 'DOI'}&.fetch('identifier') || ''
```

Reply:
I used filter_map here, and made the other method changes